crates/ruff/src/rules/pycodestyle/rules/too_many_blank_lines.rs

```diff
@@ -0,0 +1,92 @@
+use ruff_python_ast::source_code::Locator;
+use ruff_text_size::{TextRange, TextSize};
+
+use ruff_diagnostics::{AlwaysAutofixableViolation, Diagnostic, Edit, Fix};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::newlines::Line;
+
+/// ## What it does
+/// Checks for too many (>=3) blank lines.
+///
+/// ## Why is this bad?
+/// PEP 8 recommends the using blank lines as following:
+/// - Two blank lines are expected between functions and classes
+/// - One blank line is expected between methods of a class.
+///
+/// ## Example
+/// ```python
+/// def func1():
+///     pass
+///
+///
+///
+/// def func2():
+///     pass
+/// ```
+///
+/// Use instead:
+/// ```python
+/// def func1():
+///     pass
+///
+///
+/// def func2():
+///     pass
+/// ```
+///
+/// ## References
+/// - [PEP 8](https://peps.python.org/pep-0008/#blank-lines)
+#[violation]
+pub struct TooManyBlankLines(pub usize);
+
+impl AlwaysAutofixableViolation for TooManyBlankLines {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let TooManyBlankLines(nb_blank_lines) = self;
+        format!("Too many blank lines ({nb_blank_lines})")
+    }
+
+    fn autofix_title(&self) -> String {
+        "Remove extraneous blank line(s)".to_string()
+    }
+}
+
+/// E303
+pub(crate) fn too_many_blank_lines(line: &Line, locator: &Locator) -> Option<Diagnostic> {
+    // Only check for too many blank lines starting from the first blank line of a (potential) series
+    // of blank lines (to avoid duplicate errors).
+    // Also ignore blank lines at the beginning of the file.
+    if line.start().to_u32() > 0
+        && line.trim().is_empty()
+        && !locator
+            .line(TextSize::new(line.start().to_u32() - 1))
+            .trim()
+            .is_empty()
```

Comment:
Determining the line is fairly expensive using. You may also be unlucky if the previous line ends with `\r\n` because you then index right between the `\r` and `\n` and there's no good way for the locator to know that it is now between a `\r\n`. 

I'm not quite sure what the best way is to implement this rule. Ideally, it could keep a state between each line, counting the empty lines to this point (CC: @charliermarsh).

Reply:
I believe pycodestyle tracks the number of empty lines while building up logical lines -- we should probably do the same? That's now straightforward because we emit `NonLogicalNewline` whenever we see an empty line.

Reply:
You'd place it at the next logical line -- you'd check the number of preceding empty lines, I think?

Reply:
Looking at the pydocstyle rule again, my implementation is wrong anyway. The following code should generate an error:

```python
class MyClass(object):
    def func1():
        pass


    def func2():
        pass
```
-> `E303 too many blank lines (2)`


I'll try again using logical lines.

Reply:
@MichaReiser @charliermarsh I've had another go using logical lines. It's a very, very rough draft (that might have broken a few other rules), but could you tell me if you think it's worth continuing on that path ?

The modifications I made to the `logical_lines.rs` file can be seen [here](https://github.com/charliermarsh/ruff/compare/main...hoel-bagard:ruff:add_E301?expand=1#diff-c44781a9e0e9d6258049aeaaf314a3e415ee3ffd19e7a32cb6a00945f7cb219d), and the checking function is [here](https://github.com/hoel-bagard/ruff/blob/add_E301/crates/ruff/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs#L300). I've also had to remove the explicit skip of empty lines [here](https://github.com/charliermarsh/ruff/compare/main...hoel-bagard:ruff:add_E301?expand=1#diff-dc30d96352cd482de3acfae061d6c936c886fb1b98a00d84e9ae4dbc057e11f8R153).

I still haven't solved the issue with `\r\n`, but maybe I could count the number of characters to remove along with the number of blank lines ?